File: BOJ/Silver/1244.js
// 스위치 켜고 끄기

const fs = require('fs');
const filePath = process.platform === 'linux' ? '/dev/stdin' : '../input.txt';
const input = fs.readFileSync(filePath).toString().split('\n');

let switchNum = Number(input[0]);
let state = input[1].split(' ');
let studentNum = Number(input[2]);
let studentInfo = [];

for(let i = 3; i<3+studentNum; i++) {
    let [gender , have] = input[i].split(' ');
    studentInfo.push({gender,have});
}


for(let i = 0; i<studentNum; i++) {
    let cur_student_gender = parseInt(studentInfo[i].gender);
    let cur_student_have = parseInt(studentInfo[i].have);
    if(cur_student_gender === 1) {
        for(let j = 0; j<switchNum; j++) { // 남학생
            if((j+1)%cur_student_have == 0) { // 스위치 번호가 자기가 받은 수의 배수인 경우   
                state[j] = parseInt(state[j]) == 0 ? 1 : 0; // 상태 변경 
            }
        }
    }else {
        changeState(cur_student_have-1);
    }
} 


function changeState(have) {    // 여학생의 경우
    state[have] = (state[have]) == 0 ? 1 : 0;
    
    for(let i = 1; i<=switchNum/2; i++) {
        if((state[have-i]) == (state[have+i])) {
            state[have-i] = (state[have-i]) == 0 ? 1 : 0;
            state[have+i] = (state[have+i]) == 0 ? 1 : 0;
        }else {
            break;
        }
    } 
    
}

let answer = '';
for(let i = 0; i<state.length; i++) {
    answer += `${state[i]} `;
    if((i+1) % 20 == 0) {
        answer += '\n'
    }
}

console.log(answer);

File: BOJ/Silver/1244.py
n = int(input())

switch_arr = list(map(int,input().split()))

student_num = int(input())

def switch_state(gender, number, switches):
  switch_count = len(switches)
  # 남학생인 경우
  if gender == 1:
    change_num = number
    while change_num <= switch_count :
      idx_to_change = change_num - 1
      switches[idx_to_change] = 1 - switches[idx_to_change]
      change_num += number
  # 여학생인 경우
  elif gender == 2 :
    center_idx = number - 1
    switches[center_idx] = 1 - switches[center_idx]
    offset = 1
    while True:
      left_idx = center_idx - offset
      right_idx = center_idx + offset
      if left_idx < 0 or right_idx >= switch_count:
        break
      if switches[left_idx] == switches[right_idx]:
        switches[left_idx] = 1 - switches[left_idx]
        switches[right_idx] = 1 - switches[right_idx]
        offset += 1
      else: 
        break

for _ in range(student_num):
  gender, switch_num = map(int,input().split())
  switch_state(gender,switch_num, switch_arr)

for i in range(n):
  print(switch_arr[i], end=" ")
  if (i+1)%20 == 0:
    print()